File: src/main/java/me/fourteendoggo/xkingdoms/commands/vanish/VanishManager.java
package me.fourteendoggo.xkingdoms.commands.vanish;

import me.fourteendoggo.xkingdoms.XKingdoms;
import me.fourteendoggo.xkingdoms.lang.LangKey;
import me.fourteendoggo.xkingdoms.utils.Constants;
import me.fourteendoggo.xkingdoms.utils.Reloadable;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BarStyle;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import java.util.UUID;

public class VanishManager implements Reloadable {
    /**
     * Player is not vanished but has vanish flag set
     * This means that the player has been vanishing before and is now back visible
     */
    private static final byte NOT_VANISHED = 0;
    /**
     * Player is vanished and the config stated that we should give them night vision
     */
    private static final byte VANISHED_WITH_NIGHT_VISION = 1;
    /**
     * Player is vanished and we didn't give them night vision
     */
    private static final byte VANISHED_WITHOUT_NIGHT_VISION = 2;
    private final XKingdoms plugin;
    private final NamespacedKey key;
    private final BossBar bossBar;
    private final Set<UUID> vanished;

    public VanishManager(XKingdoms plugin) {
        this.plugin = plugin;
        this.key = new NamespacedKey(plugin, "vanished");
        this.bossBar = Bukkit.createBossBar("Vanished", getBossBarColor(), BarStyle.SOLID);
        this.vanished = new HashSet<>();

        bossBar.removeAll(); // just to be sure
    }

    @Override
    public void reload() {
        BarColor color = getBossBarColor();
        bossBar.setColor(color);
    }

    private BarColor getBossBarColor() {
        String color = plugin.getConfig().getString("vanish.bossbar-color");
        if (color == null || color.isBlank()) {
            return BarColor.PURPLE;
        }
        return switch (color.toLowerCase()) {
            case "pink" -> BarColor.PINK;
            case "blue" -> BarColor.BLUE;
            case "red" -> BarColor.RED;
            case "green" -> BarColor.GREEN;
            case "yellow" -> BarColor.YELLOW;
            case "white" -> BarColor.WHITE;
            default -> BarColor.PURPLE;
        };
    }

    public void destroyState() {
        bossBar.removeAll(); // doing this ensures that boss bars do not appear twice after reloading
    }

    private boolean isValidVanishFlag(byte flag) {
        return flag == VANISHED_WITH_NIGHT_VISION || flag == VANISHED_WITHOUT_NIGHT_VISION;
    }

    public Iterator<UUID> getVanishedPlayers() {
        return vanished.iterator();
    }

    public boolean vanishIfFlagged(Player target) {
        PersistentDataContainer pdc = target.getPersistentDataContainer();
        byte flag = pdc.getOrDefault(key, PersistentDataType.BYTE, (byte) -1);

        if (!isValidVanishFlag(flag)) {
            return false;
        }
        return vanish(target, true, false, flag);
    }

    public boolean isVanished(Player player) {
        return vanished.contains(player.getUniqueId());
    }

    public void toggle(Player target) {
        if (unvanish(target, true, true)) {
            target.sendMessage(plugin.getLang(LangKey.VANISH_DISABLED));
        } else {
            vanish(target, false); // false cuz unvanish() already checked and returned false
            target.sendMessage(plugin.getLang(LangKey.VANISH_ENABLED));
        }
    }

    public boolean vanish(Player target, boolean check) {
        byte vanishFlag = plugin.getConfig().getBoolean("vanish.apply-night-vision")
                ? VANISHED_WITH_NIGHT_VISION
                : VANISHED_WITHOUT_NIGHT_VISION;
        return vanish(target, check, true, vanishFlag);
    }

    public boolean vanish(Player target, boolean check, boolean notifyStaff, byte vanishFlag) {
        if (!vanished.add(target.getUniqueId()) && check) { // ensure set has a valid state
            return false;
        }

        for (Player player : Bukkit.getOnlinePlayers()) {
            if (player == target || !player.canSee(target)) continue;
            player.hidePlayer(plugin, target);

            if (notifyStaff && !player.hasPermission(Constants.MODERATOR_PERMISSION_STRING)) continue;
            player.sendMessage(plugin.getLang(LangKey.VANISH_ANNOUNCE, target.getName()));
        }
        changePlayerState(target, true);

        if (vanishFlag == VANISHED_WITH_NIGHT_VISION) {
            target.addPotionEffect(new PotionEffect(PotionEffectType.NIGHT_VISION, Integer.MAX_VALUE, 1, false, false, false));
        }
        target.getPersistentDataContainer().set(key, PersistentDataType.BYTE, vanishFlag);

        bossBar.addPlayer(target);
        target.setSaturation(20);
        target.setExhaustion(0);
        return true;
    }

    public boolean unvanish(Player target, boolean check, boolean makePersistent) {
        if (!vanished.remove(target.getUniqueId()) && check) { // ensure set has a valid state
            return false;
        }

        for (Player player : Bukkit.getOnlinePlayers()) {
            if (player == target || !player.canSee(target)) continue;
            if (!player.hasPermission(Constants.MODERATOR_PERMISSION_STRING)) continue;

            player.sendMessage(plugin.getLang(LangKey.VANISH_BACK_VISIBLE_ANNOUNCE, target.getName()));
        }
        changePlayerState(target, false);
        bossBar.removePlayer(target);

        PersistentDataContainer pdc = target.getPersistentDataContainer();
        byte vanishFlag = pdc.getOrDefault(key, PersistentDataType.BYTE, (byte) -1);

        if (vanishFlag == VANISHED_WITHOUT_NIGHT_VISION) {
            target.removePotionEffect(PotionEffectType.NIGHT_VISION);
        }
        if (makePersistent) {
            pdc.set(key, PersistentDataType.BYTE, NOT_VANISHED);
        }
        return true;
    }

    private void changePlayerState(Player target, boolean vanish) {
        target.setInvulnerable(vanish);
        target.setSleepingIgnored(vanish);

        target.setAllowFlight(vanish || switch (target.getGameMode()) {
            case CREATIVE, SPECTATOR -> true;
            default -> false;
        });
    }
}

File: src/main/java/me/fourteendoggo/xkingdoms/commands/vanish/VanishManager.kt
package me.fourteendoggo.xkingdoms.commands.vanish

import me.fourteendoggo.xkingdoms.XKingdoms
import me.fourteendoggo.xkingdoms.lang.LangKey
import me.fourteendoggo.xkingdoms.utils.Constants
import me.fourteendoggo.xkingdoms.utils.Reloadable
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.NamespacedKey
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import org.bukkit.boss.BossBar
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataType
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import java.util.*

class VanishManager(private val plugin: XKingdoms) : Reloadable {
    private val key = NamespacedKey(plugin, "vanished")
    private val bossBar: BossBar
    private val vanished = HashSet<UUID>()

    init {
        bossBar = Bukkit.createBossBar("Vanished", bossBarColor, BarStyle.SOLID)
        bossBar.removeAll() // just to be sure
    }

    override fun reload() {
        bossBar.color = bossBarColor
    }

    private val bossBarColor: BarColor
        get() {
            val color = plugin.config.getString("vanish.bossbar-color")
            return if (color.isNullOrBlank()) {
                BarColor.PURPLE
            } else when (color.lowercase()) {
                "pink" -> BarColor.PINK
                "blue" -> BarColor.BLUE
                "red" -> BarColor.RED
                "green" -> BarColor.GREEN
                "yellow" -> BarColor.YELLOW
                "white" -> BarColor.WHITE
                else -> BarColor.PURPLE
            }
        }

    fun destroyState() = bossBar.removeAll() // doing this ensures that boss bars do not appear twice after reloading

    private fun isValidVanishFlag(flag: Byte): Boolean {
        return flag == VANISHED_WITH_NIGHT_VISION || flag == VANISHED_WITHOUT_NIGHT_VISION
    }

    val vanishedPlayers: Iterator<UUID>
        get() = vanished.iterator()

    fun vanishIfFlagged(target: Player): Boolean {
        val pdc = target.persistentDataContainer
        val flag = pdc.getOrDefault(key, PersistentDataType.BYTE, -1).toByte()
        return if (!isValidVanishFlag(flag)) {
            false
        } else vanish(target, check = true, notifyStaff = false, flag)
    }

    fun isVanished(player: Player): Boolean {
        return vanished.contains(player.uniqueId)
    }

    fun toggle(target: Player) {
        if (unvanish(target, check = true, makePersistent = true)) {
            target.sendMessage(plugin.getLang(LangKey.VANISH_DISABLED))
        } else {
            vanish(target, false) // false because unvanish() already checked and returned false
            target.sendMessage(plugin.getLang(LangKey.VANISH_ENABLED))
        }
    }

    fun vanish(target: Player, check: Boolean): Boolean {
        val vanishFlag = when (plugin.config.getBoolean("vanish.apply-night-vision")) {
            true -> VANISHED_WITH_NIGHT_VISION
            false -> VANISHED_WITHOUT_NIGHT_VISION
        }
        return vanish(target, check, true, vanishFlag)
    }

    private fun vanish(target: Player, check: Boolean, notifyStaff: Boolean, vanishFlag: Byte): Boolean {
        if (!vanished.add(target.uniqueId) && check) { // ensure set has a valid state
            return false
        }
        for (player in Bukkit.getOnlinePlayers()) {
            if (player === target || !player.canSee(target)) continue
            player.hidePlayer(plugin, target)
            if (notifyStaff && !player.hasPermission(Constants.MODERATOR_PERMISSION_STRING)) continue
            player.sendMessage(plugin.getLang(LangKey.VANISH_ANNOUNCE, target.name))
        }
        changePlayerState(target, true)
        if (vanishFlag == VANISHED_WITH_NIGHT_VISION) {
            target.addPotionEffect(PotionEffect(PotionEffectType.NIGHT_VISION, Int.MAX_VALUE, 1, false, false, false))
        }
        target.persistentDataContainer.set(key, PersistentDataType.BYTE, vanishFlag)
        bossBar.addPlayer(target)
        target.saturation = 20f
        target.exhaustion = 0f
        return true
    }

    fun unvanish(target: Player, check: Boolean, makePersistent: Boolean): Boolean {
        if (!vanished.remove(target.uniqueId) && check) { // ensure set has a valid state
            return false
        }
        for (player in Bukkit.getOnlinePlayers()) {
            if (player === target || !player.canSee(target)) continue
            if (!player.hasPermission(Constants.MODERATOR_PERMISSION_STRING)) continue
            player.sendMessage(plugin.getLang(LangKey.VANISH_BACK_VISIBLE_ANNOUNCE, target.name))
        }
        changePlayerState(target, false)
        bossBar.removePlayer(target)
        val pdc = target.persistentDataContainer
        val vanishFlag = pdc.getOrDefault(key, PersistentDataType.BYTE, -1).toByte()
        if (vanishFlag == VANISHED_WITHOUT_NIGHT_VISION) {
            target.removePotionEffect(PotionEffectType.NIGHT_VISION)
        }
        if (makePersistent) {
            pdc.set(key, PersistentDataType.BYTE, NOT_VANISHED)
        }
        return true
    }

    private fun changePlayerState(target: Player, vanish: Boolean) {
        target.isInvulnerable = vanish
        target.isSleepingIgnored = vanish
        target.allowFlight = vanish || when (target.gameMode) {
            GameMode.CREATIVE, GameMode.SPECTATOR -> true
            else -> false
        }
    }

    companion object {
        /**
         * Player is not vanished but has vanish flag set
         * This means that the player has been vanishing before and is now back visible
         */
        private const val NOT_VANISHED: Byte = 0

        /**
         * Player is vanished and the config stated that we should give them night vision
         */
        private const val VANISHED_WITH_NIGHT_VISION: Byte = 1

        /**
         * Player is vanished and we didn't give them night vision
         */
        private const val VANISHED_WITHOUT_NIGHT_VISION: Byte = 2
    }
}
